File: commands/stock.py
#!/usr/bin/env python3
import os
import requests
import discord


async def stock(ctx, symbol):

    embed = discord.Embed(color=0xfee9b6, title="⏳  Loading...")
    interaction = await ctx.respond(embed=embed)

    STOCK_KEY = str(os.getenv('STOCK_KEY'))

    r = requests.get(f'https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol={symbol}&apikey={STOCK_KEY}').json()

    symbol = r["Global Quote"]["01. symbol"]
    price = str(round(float(r["Global Quote"]["05. price"]), 2))
    change = str(round(float(r["Global Quote"]["09. change"]), 2))

    emoji = "📈"
    color = 0x77b354

    if "-" in change:
        emoji = "📉"
        color = 0xdd2f45


    content = f"The current stock price of {symbol} is ${price}"
    content += f"\n The price has changed by ${change} today  {emoji}"

    embed = discord.Embed(color=color,
                          title=f"{emoji}  Stock Price of `{symbol}` is ${price}",
                          description=f"The price has changed by ${change} today.")

    await interaction.edit_original_response(embed=embed)

File: commands/stock.ts
import {
  DiscordInteraction,
  InteractionResponse,
  Colors,
} from "../types/discord";
import { embedResponse, errorResponse } from "../utils/discord";
import { getOptionValue } from "../handlers/commands";

export async function handleStock(
  interaction: DiscordInteraction,
): Promise<InteractionResponse> {
  const symbol = getOptionValue<string>(interaction, "symbol")!.toUpperCase();
  const apiKey = process.env.STOCK_KEY;

  if (!apiKey) {
    return errorResponse("Stock API not configured.");
  }

  try {
    const url = `https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${symbol}&apikey=${apiKey}`;
    const response = await fetch(url);
    const data = (await response.json()) as {
      "Global Quote": Record<string, string>;
    };

    const quote = data["Global Quote"];
    if (!quote || !quote["01. symbol"]) {
      return errorResponse(`Could not find stock data for ${symbol}.`);
    }

    const stockSymbol = quote["01. symbol"];
    const price = parseFloat(quote["05. price"]).toFixed(2);
    const change = parseFloat(quote["09. change"]).toFixed(2);
    const changePercent = quote["10. change percent"];

    const isPositive = !change.startsWith("-");
    const emoji = isPositive ? "📈" : "📉";
    const color = isPositive ? Colors.GREEN : Colors.RED;

    return embedResponse({
      title: `${emoji} Stock Price of \`${stockSymbol}\` is $${price}`,
      description: `The price has changed by $${change} (${changePercent}) today.`,
      color,
    });
  } catch (error) {
    console.error("Stock API error:", error);
    return errorResponse("Failed to fetch stock data. Please try again later.");
  }
}
